Conflict 1.0

TYWong99 opened the pull request:
Added CLI & logic module for scanning for conflicts within each sort_layer.

Other changes:
 - (level_playdo) When obtaining objects w/ names, now also search within folders 1-layer deep
 - (tiled_utils) Shortcut functions for obtaining object properties
 - (cli_test) Edited to better function as a CLI template

TYWong99 committed:
Conflict 1.0

Added CLI & logic module for scanning for conflicts within each sort_layer.

Other changes:
 - (level_playdo) When obtaining objects w/ names, now also search within folders 1-layer deep
 - (tiled_utils) Shortcut functions for obtaining object properties
 - (cli_test) Edited to better function as a CLI template

qtran122 commented on CLI_sort_conflict.py:
```diff
@@ -0,0 +1,74 @@
+
+    
+USAGE EXAMPLE:
+    cd /Users/Jimmy/20-GitHub/StarTools
+    clear; python CLI_sort_conflict.py j04 --v 2
```
remember to keep the Usage example simple and clean. The way I have it setup, I don't ever need to cd into any folder. Also, why 'clear'?

Simple "python cli_sort_conflict.py j04 --v 2"

TYWong99 replied:
The command is for clearing all previous outputs from the terminal window. You may recall in the Unity Editor of how each time your run the game, you can toggle the option to clear all previous outputs. The clear command is for my window to not be cluttered up, and know this output is from this session only.

This is one of the "debugging" stuff that I do each time I start a new project, and will remove them only near the end when I've already finished most of the testing. Unless you're really against keeping them here temporarily, I think I would leave the cd & clear here for now.

qtran122 replied:
This is silly. You can keep "cd" and "clear" in your personal files. ANY workplace, when you submit code for review - everything that is in that review is fair game for critique. To have chunks of code in there that aren't actually there and have no indication that they are temporary is bad code hygiene, period.

Consider it from my perspective - how am I supposed to know what's fair to critique and what not to critique? I'm not a mind reader. Please remove

qtran122 commented on CLI_sort_conflict.py:
```diff
@@ -0,0 +1,74 @@
+'''
```
this file's name is CLI_sort_conflict.py

For consistency with other cli's, let's use lowercase: cli_sort_conflict.py

TYWong99 replied:
Oh I thought you were trying to make a new convention since the suggested name back then uses capital letters. I'll change it back to small letters next time.

qtran122 commented on CLI_sort_conflict.py:
```diff
@@ -0,0 +1,74 @@
+#--------------------------------------------------#
+'''Main'''
+
+arg_description = 'Process a tiled level XML and ...'
```
complete description

qtran122 commented on logic/common/level_playdo.py:
```diff
@@ -125,7 +125,14 @@ def DuplicateObjectGroup(self, object_group_to_copy, new_object_group_name, attr
+        # Can search in object_layer outside fodlers and 1 layer of folders
+        obj_layer_list = self.level_root.findall('objectgroup')
+        for folder in self.level_root.findall('group'):
+            obj_layer_list += folder.findall('objectgroup')
```
GJ! You identified a weakness in this function (its inability to navigate folders) and found an answer that works! Definitely a passing grade 👍 

Though querying something, and then adding to the query results after the fact is a little awkward 🤔 Usually when you query something, it's one nice clean query. To have to do more means that our query is malformed. It was a sign to me that we're probably navigating the elementtree the wrong way, and to step back and research the canonical solution. Turns out there's a way to do this even more cleanly than before:

Turns out that any library that deals with XML also implements XPath (XML Path Language) - which is an agreed upon standard for how to navigate XML's. One of the most common XPath expressions is  `.//element` - this will select all elements named "element" at any depth starting from the current node.

So with that, here's a more elegant approach to GetAllObjectsWithName

```
def GetAllObjectsWithName(self, object_name):
        '''Searches all object groups to find objects with the given object_name'''
        objects_w_matching_names = []
        for obj in self.level_root.findall(".//object"):
            if obj.get("name") == object_name:
                objects_w_matching_names.append(obj)
    
        return objects_w_matching_names
```

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -164,6 +164,22 @@ def RotateTileId(tile_id):
 #--------------------------------------------------#
 '''Object Properties'''
 
+def GetProperty( curr_obj, property_name ):
```
I think housing this function inside level_playdo.py would be a better fit.

None of the functions inside tiled_utils.py deal with an ElementTree object, but there's precedent for that inside level_playdo.py. See the function "_AddPropertiesToObject". Putting this function right above that one would be a good place.

Also, let's rename the function to "GetPropertyFromObject"

TYWong99 replied:
I actually think it's best NOT to put it inside `level_playdo.py`. If we are to place it next to `_AddPropertiesToObject`, i.e. outside the class definition, it would not be callable without importing the playdo file, which we normally don't outside of CLI.

But if we are to place the function inside class definition, it would then become inaccessible in one of the functions in `conflict.py`, where the playdo is not passed down. I'm sure there might be more instances of that in the future too, and having to pass down playdo each time just feels like an inelegant solution?

Overall, I think keeping this function in one of the "utility files" is the simplest solution, since housing publicly accessible utility functions is exactly what they're for. In contrast, `playdo.py` is a "class file", which I believe should contain the class as the only publicly accessible thing.

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -164,6 +164,22 @@ def RotateTileId(tile_id):
 
+def GetProperty( curr_obj, property_name ):
+    '''Extract the property as string, e.g. returns 'BRONZE' from GetProperty('_type')'''
+    for curr_property in curr_obj.find('properties').findall('property'):
```
The use of the variable name "curr_obj" is a bit odd. Usually when you prepend "curr_" to something, that's because there's a need to distinguish the current object from the "previous" object or the "next" object (like when you do a sort problem). But here, there is no need for that extra bit of info.

For example, imagine if I was introducing you to someone, and I said "This is Ty. He's my 6-foot tall assistant" - does an adjective in there feel a little extraneous? (6-foot)

You can use "obj" for name. Unlike "str", obj is not a reserved or in-built type in Python.

But I'd prefer the name "tiled_object" to have consistency with the function "_AddPropertiesToObject" inside level_playdo.py

TYWong99 replied:
Renamed.

Also sadly I am not 6 foot.

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -164,6 +164,22 @@ def RotateTileId(tile_id):
+    return ''    # returning None would crash when attempted to be converted into string
+
+
+def GetSort( curr_obj ):
+    '''Take an object and extract its sort value as string, e.g. "fg_tiles,-1"'''
+    return GetProperty(curr_obj, '_sort')
```
This function doesn't have enough content to justify its existence me thinks

TYWong99 replied:
I personally prefer being lazy whenever possible, since future me may not remember how the inner workings well enough to put all the arguments correctly blind.

The 1st benefit is that the input is simple and future-proof - You have 1 input argument, and it returns 1 string afterwards. Without it, we would have to remember (1) the function name, & (2) the in-editor keyword for searching. I almost always avoid using literal constant for string (i.e. '_sort'), unless it exists only within the current file, or is used in a common module. While not likely, if we are to change the property name from '_sort' to something else like _sort_value, we would have to check every single code file to change it (instead of only changing 1 string in this GetSort() function).

The 2nd benefit is that it's what I call a "shortcut function". It serves one simple purpose, is easily understood since it's mostly just the same function call, and is easy to used like I mentioned in the previous paragraph. Besides being convenient to use, it doesn't stem too far from the GetProperty() function and thus I don't think it would hurt the overall structure by causing confusion either.

The 3rd benefit is that its name is short and intuitive (thus easily memorised)... That's it. Again, usually I just want to be lazy and type as little as I can if possible.

... So yeah, future-proof & convenient. But if you're really against it...

qtran122 replied:
I could understand your reasoning more if this was a function that was used a lot - but it's used once.

How about for now, GetSort can exist, but it resides inside sort_conflict.py until such a time comes where its existence is proven to be useful in other tools as well?

TYWong99 replied:
For personal projects, I would probably keep it there, just so future me would know to put more lazy shortcuts there whenever I feel like it, e.g. `GetName` and `GetType`.

But since it's your project, I'm just gonna remove it entirely. And like you said, it's only used once anyway.

TYWong99 replied:
Okay I forgot what the function for getting the name of an object should be, so I now insist on adding these back in for future-me.

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -172,6 +188,16 @@ def GetPolyPoints( input_string ):
 
+
+# TODO provide more info
+def GetObjPrintData( curr_obj ):
+    '''General summary for the content of an object'''
+    print_str = ""
+    print_str += curr_obj.get('name')
+    return print_str
+
```
Is this function used anywhere?

If the function is likely to just be used by conflict.py, then it should be housed in conflict.py

TYWong99 replied:
Ah this was a leftover when I was planning to make a "universally usable" function for printing object's info. But later down the line I realise most of what I'm printing would be unique to cli_conflict anyway, so I moved everything there instead.

Maybe I'll turn it into a "template" function, such that if we need to print object info in other tool, we can copy this function and add things as we see fit?

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+'''
+<SUMMARY TBA>
```
Time to add that summary!

Also, can you change the name to "sort_conflict.py". Just "conflict" by itself doesn't seem descriptive enough. A lot of things can conflict, not just the sort layers

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+'''Public Functions'''
+
+def CheckConflicts(playdo, list_scan_obj):
+	''' Task 1 - Check if there is any conflict '''
```
This function warrants a more detailed description with an explanation of the args and return value. See for example "CreateTilesCollage" in image_utils.py

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+		log.Info( f'# of \"{scanned_obj}\": {len(curr_obj_list)}' )
+		for obj in curr_obj_list:
+			curr_string = tiled_utils.GetSort(obj)
+			if curr_string == None: continue
```
would this ever be None? GetSort, which is a wrapper around GetProperty seems to always at least return an empty string?

TYWong99 replied:
I actually was debating whether GetProperty should return None of an empty string. I know the "proper" way is to return None since there's nothing to return, and there *should* be a distinction from property simply does not need a value (e.g. `autoset` for BB).

So, returning a empty string is more like a band-aid fix, one that I want to hear if you have alternatives available, whereas the "continue if None is returned" here feels like the proper way to do things.

TYWong99 replied:
As mentioned, I anticipate None to be for when the object does not contain the property, and an empty string to be for when the object does contain the property but with empty string as value. Would adding the comment plus an extra if-statement be more preferable?

```
if curr_string == None: continue  # When object does not contain `_sort`
if curr_string == '': continue    # When property has property but no value
```

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+USAGE EXAMPLE:
+    raw_dict = conflict.CheckConflicts(playdo, _LIST_LIGHTING_OBJ)
+    pruned_dict = conflict.PruneConflicts(playdo, conflict_dictionary)
+    conflict.FixConflicts(playdo, pruned_dict)
```
Nice! I like this setup and organization.

I'd add an additional step, "PrintPotentialConflicts", and move all the printing into that function and out of of CheckConflicts.

I'd also rename some functions:
Instead of "raw_dict". How about "sortval_to_objects_map"?
Instead of "CheckConflicts". How about "OrganizeObjectsBySortVal"?

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+	log.Info("\n--- Checking for sort value conflicts ---\n")
+
+	# Create the dictionary with each unique sort value as keys
+	dict = {}
```
instead of dict, let's go with "sortval_to_objects_map"

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+	max_name_len = len( max(list_scan_obj, key = len) )    # Length of longest name in list
+	log.Info('')
+	log.Info('~~Note: Sort Layers with only 1 element are skipped~~')
+	for key, value in dict.items():
```
never use "key" and "value" for iterating through a dictionary, always better to name what the key and value are. In an interview, the interviewer would be looking for this

Also, when you display the results, can you go about it in a "sorted" fashion? So display the fg_tiles with the highest sort numbers first. Then the game_objects, then the bg_tiles. For example:

fg_tiles,300
fg_tiles,52
game_objects,100
game_objects,-1
bg_tiles,200
bg_tiles,-5

TYWong99 replied:
Name changed from `key` to `key_sort`, also `value` to `list_obj`.

Also you mentioned not needing to spend too much effort on the output message later since you likely wouldn't be looking at it after all. Want to confirm whether that's true, or whether you still want me to sort the itmes.

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+	return ( s + ' ' * (min_len-len(s)) )
+
+
+def s2tu(num_in_str):
```
Let's expand the name for readability

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+#	return str( int(s2i(num_in_str)) / 16 )
+
+
+def s2i(num_in_str):
```
expand name for readability

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+#--------------------------------------------------#
+'''General Utility, to be relocated?'''
+
+def Indent(s, min_len):
```
I'd prefer keeping these functions specific to this file. Let's call it "_Indent" (Having an underbar in front was my old work place's methodology for signifying private functions)

TYWong99 replied:
I vaguely remember reading somewhere, that people use 1 underscore for "protected" functions/variables, and 2 underscores for "private" instead. :thinking emoji:

qtran122 replied:
I did say that it was my old work place's methodology - so consider it a personal preference.

My reasoning is we don't use any inheritance, so we would never have cases where we get to use one underbar in that scheme.

Thinking about it, I haven't used inheritance for years due to Unity's entity component system way of doing things. And Interfaces can take care of the shared functionality between classes in a more flexible way... so yeah, just haven't needed inheritance.

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+    num = str(num)[:str(num).index('.')+dec+2]
+    if num[-1]>='5':
+        return float(num[:-2-(not dec)]+str(int(num[-2-(not dec)])+1))
+    return float(num[:-1])
```
What does this function have over Python's built-in "round" function?

The link you posted also seems to say to use "round" ?

(on stackoverflow, look at the number of upvotes to see the topranked answer)

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+	log.Info('')
+	log.Info('~~Note: Sort Layers with only 1 element are skipped~~')
+	for key, value in dict.items():
+		if len(value) == 1: continue
```
Can you make it so that if --v=2, even the layers with just one element will be listed? They're actually crucial to see for when we go to "fix" conflicts later.

TYWong99 replied:
Not sure what this have to do with verbosity, but are you saying to simply not skip skip sort layers with only 1 element?

qtran122 replied:
Hmmm... yeah, thinking on it more, there hasn't really been any data in this tool so far that I would consider extraneous.

Sort layers with only 1 element should be listed!

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+'''Utility'''
+
+
+def GetParentName(obj, playdo):
+	parent = GetParent(obj, playdo)
+	if parent == None: return '...'
+	return parent.get('name')[8:]    # Remove beginning, i.e. 'objects_'
+
+
+# TODO relocate to tiled_utils?
+# TODO List comprehension?
+def GetParent(obj, playdo):
+	'''Return the name of the layer that the object is in'''
+	root = playdo.level_root
+
+	# Check all object layers that are in 0~1 layer of folder
+	all_objectgroup = root.findall('objectgroup')
+	for folder in root.findall('group'):
+		all_objectgroup += folder.findall('objectgroup')
+
+	# Check which objectgroup contains the obj
+	for group in all_objectgroup:
+		for child in group:
+			if child == obj: return group
+	return None
```
Pretty good attempt! It definitely gets the job done.

Though it's sorta wasteful with how much we're iterating to re-get each parent. Looking online, I found this which does the trick nicely:
https://stackoverflow.com/questions/2170610/access-elementtree-node-parent-node

It's a one-liner:
`parent_map = {c: p for p in tree.iter() for c in p}`

Construct the parent_map once and cache it. After that, you can get the parent for each object with a simple dictionary fetch. Give it a shot.

If you get stuck, here's an example with some wider context on how it'd be used ( https://imgur.com/WwUISPf)

qtran122 commented on logic/standalone/conflict.py:
```diff
@@ -0,0 +1,209 @@
+			dict[curr_string].append(obj)
+
+
+	# Print the result
```
Let's move printing the result out to another function (see my comment 2nd top-most comment on this file)

qtran122 reviewed (changes requested):
GJ! Some edits will be needed, but it's a solid first foray into the sort_conflicts tool.

qtran122 reviewed (changes requested):
Keep at it 👍

TYWong99 committed:
Conflict 1.1

Addressed various requested changes. Some changes are to be discussed further, and some changes are to be dealt with next time. See also the comments here and on Trello.

Other changes:

 - Now automatically check for object layers' names for consistent indentation when printing
 - Now also prints poly_line's vertice data
 - Polyline points can be obtained directly from inputting an object

TYWong99 committed:
Printing & Logging changes

Minor change to the printing function in `sort_conflict.py`.

Different log files are now generated based on input verbosity. Uploaded the 3 logged files for debugging purposes, and remove directory for privacy/security reasons.

qtran122 commented on input/remaps/remap_all.xml:
Isn't remap_all part of another effort (the upgrade to cli_remap?)

It's fine this time, but remember to use different branches. One branch = 1 pull request = 1 main "effort"

TYWong99 replied:
It's more of a "non-important" update for me. Since you may use different `remap_all` on your end (so my file doesn't affect you much), including it in the official folder is more for convenience. I dunno how to explain, but I copy and overwrite folders fairly frequently for backing up files, so I used to have to copy the `remap_all` file every time I get back to tiling, which is a bit inconvenient.

Regardless, I'll be reworking `cli_remap` eventually, right? At which point the `remap_all` wouldn't be on GitHub anymore anyway, let's just keep the file here and ignore it for now.

qtran122 commented on logic/common/level_playdo.py:
```diff
@@ -43,8 +44,9 @@ def __init__(self, file_name):
-        log.Info(f"-- level_playdo.py : initialized {file_name} ...")
+
+#        log.Info(f"-- level_playdo.py : initialized {file_name} ...")
+        log.Info(f"-- level_playdo.py : initialized {os.path.basename(file_name)} ...")
```
Nice! This is neat. I didn't know basename existed and I think file_utils has a "strip" function which did what this is. Obviously, basename is better. GJ!

TYWong99 replied:
I also never knew about it until looking it up online, and was going to do the strip method too. Though, I think it does require importing os to our code just for using `basename()` . Glad you don't mind that.

qtran122 commented on logic/common/level_playdo.py:
```diff
@@ -123,12 +125,11 @@ def DuplicateObjectGroup(self, object_group_to_copy, new_object_group_name, attr
-            for object in object_group.findall('object'):
-                if object.get('name') == object_name:
-                    objects_w_matching_names.append(object)
+        for obj in self.level_root.findall(".//object"):
```
neat! This is a good fix.
Buuut, it was already applied in a previous code submission. Time to update!

qtran122 commented on logic/standalone/sort_conflict.py:
```diff
@@ -0,0 +1,198 @@
+	return ( s + ' ' * (min_len-len(s)) )
+
+
+def s2tu(num_in_str):
```
Give proper name please! I don't need to know the exact X to Y conversion, but I'd like to know the main point of the function at a glance. For example:
"_FormatNumS2TU"
And then the comments can tell the rest

qtran122 commented on logic/standalone/sort_conflict.py:
```diff
@@ -0,0 +1,198 @@
+#	return str( int(s2i(num_in_str)) / 16 )
+
+
+def s2i(num_in_str):
```
Same as above

qtran122 commented on logic/standalone/sort_conflict.py:
```diff
@@ -0,0 +1,198 @@
+
+
+# From: https://stackoverflow.com/questions/31818050/round-number-to-nearest-integer
+def proper_round(num, dec=0):
```
"_ProperRound"

Can you add some comments here? I'm guessing there's a special edge case reason why other functions use "round" and one case has to use ProperRound. Cases like that are elucidated with descriptive comments

TYWong99 replied:
I forgot why too, but there seems to be a difference in output, and also my function is the incorrect one... orz

Regardless, I've simplified the conversion function in general, and this rounding function is removed entirely.

qtran122 reviewed (changes requested):
Looks good! See final-ish comments. Let's make a new code submission since StarTools has received many an update since the last review

TYWong99 committed:
Conflict 1.2

Major Changes:

- Functional Prototype (w/ alternate solution)
- Applied various requested changes
- Simplified conversion functions for printing
- Copied the property-adding function from `playdo` to `tiled_utils`
- Added name-/property-getting functions with alternate names (see Trello)

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -162,7 +162,29 @@ def RotateTileId(tile_id):
+            return curr_property.get('value')
+    return ''    # returning None would crash when attempted to be converted into string
+
+def GetProperty( tiled_object, property_name ):
+    return GetPropertyFromObject(tiled_object, property_name)
```
Let's get rid of this shortcut function.
First, because it has no added utility, and adds function call overhead.
And for two, because it doesn't seem to be used anyway

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -162,7 +162,29 @@ def RotateTileId(tile_id):
+
+
+def GetPropertyFromObject( tiled_object, property_name ):
+    '''Extract the property as string, e.g. returns 'BRONZE' from GetProperty('_type')'''
```
This comment could use more clarity. I had to use grep to remember if any objects had 'BRONZE' as a property value (it was pearls).

I think "_sort" would be better as an example, since that's something most objects have

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -171,6 +193,57 @@ def GetPolyPoints( input_string ):
+    if polyline_attribute == None : return None
+    points_string = polyline_attribute.get('points')
+    return GetPolyPoints(points_string)
+    return None
```
this line is never reached

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -162,7 +162,29 @@ def RotateTileId(tile_id):
+
+
 
 def GetPolyPoints( input_string ):
```
Is this function used anywhere other than "GetPolyPointsFromObject"? If not, let's prepend "_" to the name

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -171,6 +193,57 @@ def GetPolyPoints( input_string ):
 
+# TODO overload with same name?
+def GetPolyPointsFromObject( tiled_object ):
+    '''Accept object as input instead'''
```
It seems this is the outwards public facing function, so give it a more descriptive summary. What does it take - what does it give and in what format?

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -171,6 +193,57 @@ def GetPolyPoints( input_string ):
+#--------------------------------------------------#
+''''''
+
+# Template, see sort_conflict.py for usage with Indent()
+def PrintObjInfo( curr_obj ):
+    '''Default summary for the content of an object'''
+    print_str = ""
+    print_str += curr_obj.get('name')
+    return print_str
```
sort_conflict has its own PrintObjInfo()

Furthermore, we don't use inheritance or interfaces, so there's no reason for this to be here. If you plan to copy/paste this template for future use, then the proper home for that is in your personal files. This is very similar to DEBUG info and notes - ie, not for the official code space.

qtran122 commented on logic/common/tiled_utils.py:
```diff
@@ -171,6 +193,57 @@ def GetPolyPoints( input_string ):
+#--------------------------------------------------#
+'''Set object property'''
+
+def SetProperty(tiled_object, property_name, new_value):
```
SetPropertyOnObject. See my trello comment on why we would want this distinction

qtran122 reviewed (changes requested):
send the next one in a new pull request please!

TYWong99 committed:
Tab to Space

qtran122 commented:
Just doing some cleanup. We have cli_sort now so the conflict tool is entirely deprecated
